### PR TITLE
feat(dot/rpc) implement `chain_subscribeAllHeads` RPC

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- Implement `chain_subscribeAllHeads` and `chain_unsubscribeAllHeads`
- Create channels and register them into `BlockAPI` and notify the websocket every new block or finalization

## Tests

```
go test -run ^TestSubscribeAllHeads$ github.com/ChainSafe/gossamer/dot/rpc/subscription
```

## Issues

- Closes to #1741 

## Primary Reviewer 

- @timwu20

### CODE REVIEW

File: dot/rpc/subscription/listeners.go [hunk 1]
```diff
@@ -213,6 +214,81 @@ func (l *BlockFinalizedListener) Stop() error {
...
+					continue
+				}
+
+				finHead, err := modules.HeaderToJSON(*fin.Header)
+				if err != nil {
+					logger.Error("failed to convert finalized block header to JSON", "error", err)
+					continue
+				}
+
+				res := newSubcriptionBaseResponseJSON()
```
Comment (timwu20) [inline dot/rpc/subscription/listeners.go:257]: can you change this constructor to just take `Method` and `Params`.  Seems odd that we need to populate it ourselves.

File: dot/rpc/subscription/websocket.go [hunk 2]
```diff
@@ -278,6 +278,42 @@ func (c *WSConn) initBlockFinalizedListener(reqID float64, _ interface{}) (Liste
...
 }
 
+func (c *WSConn) initAllBlocksListerner(reqID float64, _ interface{}) (Listener, error) {
+	listener := new(AllBlocksListener)
+	listener.cancel = make(chan struct{})
+	listener.done = make(chan struct{})
+	listener.cancelTimeout = defaultCancelTimeout
+	listener.wsconn = c
+	listener.finalizedChan = make(chan *types.FinalisationInfo, DEFAULT_BUFFER_SIZE)
+	listener.importedChan = make(chan *types.Block, DEFAULT_BUFFER_SIZE)
```
Comment (timwu20) [inline dot/rpc/subscription/websocket.go:288]: move to constructor `newAllBlocksListener`

**Review by timwu20 — CHANGES_REQUESTED** [review]

File: dot/rpc/subscription/listeners.go (hunk 1 above)
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/listeners.go:262]: Done!

File: dot/rpc/subscription/websocket.go (hunk 2 above)
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket.go:288]: Done!

**Review by timwu20 — APPROVED** [review]

**Review by arijitAD — APPROVED** [review]

File: dot/rpc/subscription/listeners.go [hunk 3]
```diff
@@ -213,6 +214,90 @@ func (l *BlockFinalizedListener) Stop() error {
 	return cancelWithTimeout(l.cancel, l.done, l.cancelTimeout)
 }
 
+// AllBlocksListener is a listener that enables to be awere about new blocks and finalised ones
```
Comment (noot) [inline dot/rpc/subscription/listeners.go:217]:
```suggestion
// AllBlocksListener is a listener that is aware of new and newly finalised blocks```

**Pull request merged**